### PR TITLE
Add image optimizer

### PR DESCRIPTION
**Pull request by thewilkybarkid**
# ↪️ Pull Request

Adds an image optimizer using ~[Squoosh](https://squoosh.app/) or~ Sharp.

Has reasonably safe defaults, and supports a config file.

Refs https://github.com/parcel-bundler/parcel/discussions/5544, https://github.com/parcel-bundler/parcel/issues/4585, https://github.com/parcel-bundler/parcel/issues/627, https://github.com/parcel-bundler/parcel/issues/3739 and others.

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

File: packages/core/integration-tests/test/image.js [hunk 1]
```diff
@@ -88,4 +88,20 @@ describe('image', function() {
...
+      },
+    });
+
+    const imagePath = b.getBundles().find(b => b.type === 'jpg').filePath;
+
+    const buffer = await outputFS.readFile(imagePath);
+    const image = await sharp(buffer).metadata();
+
+    assert.strictEqual(image.width, 1920);
+    assert.strictEqual(image.size, 282274);
```
Comment (thewilkybarkid) [inline packages/core/integration-tests/test/image.js]: Works on my machine. 😅

File: packages/optimizers/squoosh/src/SquooshOptimizer.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+export default (new Optimizer({
+  async optimize({bundle, contents}) {
+    if (!bundle.env.shouldOptimize) {
+      return {contents};
+    }
+
+    if (!imagePool) {
+      imagePool = new ImagePool();
+    }
```
Comment (thewilkybarkid) [inline packages/optimizers/squoosh/src/SquooshOptimizer.js]: There doesn't seem to be a way to set-up/tear-down the optimizer.

**Comment by devongovett** [comment]
How does this compare with the existing image transformer, which uses sharp?

**Comment by thewilkybarkid** [comment]
Not actually checked, but it doesn't (and probably shouldn't?) set the mozjpeg option for example so would definitely be bigger.

**Comment by thewilkybarkid** [comment]
Tried out Squoosh based on https://github.com/parcel-bundler/parcel/discussions/5544#discussioncomment-252279, but after a bit of a search have found https://github.com/vercel/next.js/issues/27055 which is concerning. Maybe a Sharp optimiser would be better.

**Comment by DeMoorJasper** [comment]
Dont think sharp can really be used as an image optimizer, as far as I've tried tools like imagemin yield way better results size wise. Sharp is great for transforming images though which libraries like imagemin and squoosh only do the bare minimum of. We can definitely experiment with sharp and see how far we can take it.

**Comment by thewilkybarkid** [comment]
https://github.com/vercel/next.js/discussions/17141 has some useful comparisons (not Squoosh though).

**Comment by thewilkybarkid** [comment]
At least for that one image, using the same quality value as Squoosh's default in Sharp produces the same result. (Actually 10 bytes smaller.)

File: packages/optimizers/sharp/src/SharpOptimizer.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
+// @flow
+
+import {Optimizer} from '@parcel/plugin';
+import {blobToBuffer} from '@parcel/utils';
+import path from 'path';
+import sharp from 'sharp';
+
+const OPTIONS = {
+  jpeg: {
+    quality: 75,
```
Comment (thewilkybarkid) [inline packages/optimizers/sharp/src/SharpOptimizer.js]: To match https://github.com/GoogleChromeLabs/squoosh/blob/2bf0b904cd3ebf831c9b47e6f59ab508ae968f81/src/features/encoders/mozJPEG/shared/meta.ts#L23 (rather than proposing it as a new default).

**Comment by DeMoorJasper** [comment]
If sharp produces similar results size wise, maybe we should change some of the defaults in the image transformer and not run sharp a second time in the optimize stage, not sure how much difference this makes but I imagine it's gonna be a bit slower

**Comment by thewilkybarkid** [comment]
That assumes there was a transformation...

**Comment by thewilkybarkid** [comment]
Going to try other formats: if I understand correctly Sharp will be bigger as it doesn't use an optimiser.

The other option is imagemin, but I've only just seen that it's now marked as unmaintained so isn't suitable.

**Comment by thewilkybarkid** [comment]
Couldn't get Squoosh to reduce the image anywhere near as much as Sharp with the `palette` option.

**Comment by thewilkybarkid** [comment]
Given Squoosh is still reasonably experimental, maybe the way forward is to just use Sharp. The double pass could be removed if it's possible to know whether the transformer has already processed the image?

**Comment by thewilkybarkid** [comment]
Or just run every image through the transformer?

**Comment by DeMoorJasper** [comment]
@thewilkybarkid it's probably fine to keep it separate, also just realised we can't access shouldOptimize in the transformer so we should run the transformer for each image if we'd move this which will probably be worse than just running the optimiser.

Pretty nice how good sharp is with optimisation, didn't know that, thanks for researching all these optimisation options

File: packages/optimizers/sharp/src/SharpOptimizer.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    const ext = path.extname(bundle.name).substr(1);
+    const format = FORMATS.get(ext);
+
+    if (!format) {
+      throw new Error(`Sharp does not support ${ext} images.`);
+    }
+
+    const optimized = await sharp(await blobToBuffer(contents))
```
Comment (DeMoorJasper) [inline packages/optimizers/sharp/src/SharpOptimizer.js]: Not sure if we should add `.rotate()` to this as well to normalise the rotation metadata of images. We used to have an issue in the transformer where images would be rotated if they have a certain format, might also be caused by resizing not sure.

**Comment by devongovett** [comment]
I'm a little conflicted about whether image optimization should be opt in or not. If it's always lossless in terms of quality I guess that's ok, but if it's lossy then I think it would be potentially unexpected. Do we need some way to configure that?

**Comment by thewilkybarkid** [comment]
I was thinking of reading a Sharp config file, which has the output settings for each file type. As for defaults, could be quite conservative (i.e. Sharp's defaults).

File: packages/optimizers/sharp/src/SharpOptimizer.js (hunk 4 above)
Comment (thewilkybarkid) [reply packages/optimizers/sharp/src/SharpOptimizer.js]: Just seen #6444; looks like it shouldn't happen, but maybe worth having a test with the EXIF Orientation tag (more for the transformer though).

Comment (thewilkybarkid) [reply packages/optimizers/sharp/src/SharpOptimizer.js]: Will expand on https://github.com/parcel-bundler/parcel/pull/6647.

Comment (thewilkybarkid) [reply packages/optimizers/sharp/src/SharpOptimizer.js]: Turns out it _did_ happen, so I've solved it in https://github.com/parcel-bundler/parcel/pull/6624/commits/28fa471800b77b7e1f1e539e6c3bd1ae9a93b466.

File: packages/optimizers/sharp/src/SharpOptimizer.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    const options = {...DEFAULT_OPTIONS[format], ...config?.[format]};
+
+    const optimized = await sharp(await blobToBuffer(contents))
+      [format](options)
+      .rotate()
+      .toBuffer();
+
+    return {contents: optimized};
```
Comment (DeMoorJasper) [inline packages/optimizers/sharp/src/SharpOptimizer.js]: Should we add a check to ensure the size is smaller and return the original if that's smaller?

Comment (thewilkybarkid) [reply packages/optimizers/sharp/src/SharpOptimizer.js]: Hmm, technically you could just turn on making images progressive which would make them a bit bigger.

Comment (DeMoorJasper) [reply packages/optimizers/sharp/src/SharpOptimizer.js]: Why would you want to make them bigger? I meant just checking if the output buffer is smaller than the input buffer and return the input buffer if that is smaller as sometimes input images are already optimised a lot

File: yarn.lock [hunk 6]
```diff
@@ -2137,6 +2182,36 @@
     detect-libc "^1.0.3"
     globby "^11.0.3"
 
+"@parcel/types@2.0.0-beta.3.1":
+  version "2.0.0-beta.3.1"
+  resolved "https://registry.yarnpkg.com/@parcel/types/-/types-2.0.0-beta.3.1.tgz#0bc5255a4454c198d58403b84a66f059106e1610"
+  integrity sha512-8Cmh5/43LZ7wQeqawFMenZ/CCh+YAbyRuwBBApMCDLU3+RSpajUYyTpqa1HZVV8FggQb8E0xvPkmUbj/UHz/gw==
+
+"@parcel/utils@2.0.0-beta.3.1":
+  version "2.0.0-beta.3.1"
```
Comment (thewilkybarkid) [inline yarn.lock]: Not sure why Yarn is showing these as I bumped the versions to `rc.0` in in https://github.com/parcel-bundler/parcel/pull/6624/commits/8dfd3d7754dc0ef4a81c982d0b89f44f24074b02?

Comment (thewilkybarkid) [reply yarn.lock]:
```console
$ yarn list --pattern @parcel/utils
yarn list v1.22.10
├─ @parcel/transformer-js@2.0.0-rc.0
│  └─ self-published@2.0.0-beta.3.1
│     └─ @parcel/utils@2.0.0-beta.3.1
└─ @parcel/utils@2.0.0-rc.0
✨  Done in 1.00s.
```

Comment (thewilkybarkid) [reply yarn.lock]: Unrelated changes; happens on my `v2` branch locally too.

**Review by DeMoorJasper — APPROVED** [review]

**Comment by mischnic** [comment]
> I'm a little conflicted about whether image optimization should be opt in or not. If it's always lossless in terms of quality I guess that's ok, but if it's lossy then I think it would be potentially unexpected. Do we need some way to configure that?

So right now, it processes images by default with Sharp's default settings. Are those lossless (and guaranteed to not make the image bigger)?

**Comment by DeMoorJasper** [comment]
@mischnic it should be pretty much lossless and there's a check to fallback to the original image if the original is smaller (which is also what imagemin does)

~Just had a look at the sharp docs and by default `webp`, `avif` and `heif` are set to lossless: true, should we change those to false in our default configs?~ apparently I messed up or they changed it

Also did a small test to test the performance of sharp w/ and without compressions and compressing is about 3-4 times slower than without compression so my original idea of moving this to the transformer is a bad one

**Comment by devongovett** [comment]
I don't think it is lossless. Even decoding and re-encoding a JPEG is a lossy operation (JPEG encoding is inherently lossy). And the current defaults in this PR use indexed PNG which would definitely be lossy (only 256 colors). There _are_ lossless image compression libraries out there (e.g. jpegtran and optipng), but sharp doesn't appear to use them.

> Just had a look at the sharp docs and by default webp, avif and heif are set to lossless: true

I don't think this is true, the defaults listed [here](https://sharp.pixelplumbing.com/api-output) are false.

Again, I'm not sure this should be on by default if it's not fully lossless. Especially if there's no way to opt out for only certain images.

Perhaps we should leave this to the image transformer so users can opt-into image optimization via query params for each image. Using the `as` or `quality` query param is basically equivalent to what is in this optimizer already. We could probably enable `mozjpeg` as well, and potentially expose more options as well.

We could also have a global configuration to enable the image transformer for all images potentially. This would allow people to opt into image optimization for all images without manually adding query params to every image.

To sum up, I think it's too risky to enable lossy image optimization by default without people knowing and no way to opt-out per image. Plus, the image transformer already does most of this, and this would reduce performance because images would get processed twice.

**Comment by devongovett** [comment]
I used a couple Rust libraries to build an image optimizer that does lossless compression in https://github.com/parcel-bundler/parcel/pull/6762. Perhaps we can also offer a config file to enable the image transformer on all images to opt into lossy compression without manually adding query params. What do you think?

**Comment by thewilkybarkid** [comment]
Yeah, that would work. I understand the hesitancy over making it opt-out rather than opt-in, but some config to turn it on globally would be enough.

**Comment by devongovett** [comment]
Ok. I'm going to close this then since #6672 is merged. If you want to send a PR for a config file for the image transformer, feel free.
